### PR TITLE
Delegation of Stake

### PR DESCRIPTION
**Pull request by abhishekvispute**
# Delegation of Stake
### Fixes https://github.com/razor-network/contracts/issues/45
---

# Part 1: Delegation

### What is it

Delegating responsibility of voting in the network to existing staker, rather than running node yourself, and be part of reward mechanism.

---

### Why we need it

Many times general users want to participate in the network but don't have the resources to set up nodes. With this feature, they can participate in the network by delegating their stake to the staker, letting the staker vote on behalf of them, and being part of a reward mechanism.

---

### How we did it

In Short,

- `Staker Accepts Delegation`
Staker has an option if they want to accept delegation or not
If they do, they set up a commission which acts as a fee to the delegator

- `Delegation` 
Now delegator sees all stakers who have accepted delegation
and choose any, we would call this, **"Chosen one"** from now on
Delegators' RZR tokens would be transferred to the stakemanager. 
Chosen one's stake would be increased 
X number of sRZR would be minted and transferred to the delegator's account. 
X would depend on the current valuation of sRZR
- `Trading of sRZRs`
Now delegator can keep them or trade we are prone to that
If they trade, the receiver becomes delegator by definition

    If chosen staker is on reward positive 1 sRZR > 1 RZR, the delegator is at profit.

    If chosen staker is on reward negative 1 sRZR < 1 RZR, the delegator is at loss.

- `Withdraw`
If the delegator wants to withdraw
His sRZRs would be burned and he would get RZRs back at the current valuation.
Commission would be paid to the staker.

To be consistent 
Now even if staker stakes he would get sTokens back.

Please refer to this sheet for example

[sRZR](https://www.notion.so/b603abc874b642c4a35ae1db0447e1ce)

---

# Part 2: Lock Mechanism

(If you are not clear why we need lock on delegators, please refer to notion doc)

As we don't have any storage on delegators we can't go by how we did for stakers in the past.
even if we create storage then we would have to define separate functions for stakers and delegators.
**We needed a user agnostic lock mechanism that would work for all**
That's what Improved lock mechanism is 
We would have a new structure called Lock
and map it to [address][tokenaddress]

```jsx
struct Lock {
uint256 amount; //amount in sTokens
uint256 withdrawAfter;
}

mapping (address=> mapping(address=>Structs.Lock)) public locks; //in StakeStorage
```

 
In Unstake, the user will declare how much they want to withdraw (in sRZR)

and they wait for withdraw after period.

and then within WithDrawReleasePeriod, they withdraw, they will get RZRs back at the current valuation w.r.t sRZR, 
We burn sRZRs, update stakers stake and pass the commission to staker if withdrawal is initiated by the delegator.

Locks 2 (2)

Now what is this **"WithdrawReleasePeriod"** we didn't have this before
****Why do we this extra restriction?****

Because without it following case is possible.
Consider me as a delegator 
I would call unstake even though I don't want to withdraw.
Now I can withdraw anytime and I would get RZR back at a valuation of withdrawal call.

If the user fails to withdraw within **WithdrawReleasePeriod**, their lock expires and they have to call resetLock(), resetLock() is penalized to avoid front-run recurring unstake call. Please refer to the following for more details

**Attacks** 

1. **P+E**  
      If the attacker can use influence to signal in-opposite and influenced users are able to leave the network at that moment only without waiting to check if the attacker's signal was true.
      > Could be solved by forcing staker to participate in network even after unstake 
         We are not resistant to this with a new design as either we  can resist this or Hit and Run Before Dispute, and we prefer Hit and Run Before Dispute as it's in protocol
2. **Extra Protocol Hit and Run**

    If after unstake attacker can issue negative comments about the network and remain unaffected in withdraw.
    > Solved by releasing funds as per valuation in withdraw and not in unstake 

3. **Hit and Run Before Dispute**

    If the attacker can hit the network and leave before a dispute happens
    > Solved by locking funds and forcing staker not to participate

4. **Front-Run Unstake Call**

    If the attacker can call unstake though they don't want to withdraw and withdraw anytime after withdraw after period is passed
    > Solved by the introduction of WithDrawReleasePeriod, the user has to withdraw within this period only, so attackers cannot pre-call unstake, as if they do, after WithDrawReleasePeriod their withdrawal req becomes invalid. 

    **Front-Run Recurring Unstake call**
    To solve the previous attack we added WithDrawReleasePeriod, but with that as well one can call unstake again each time after WithDrawReleasePeriod is over, on less gas-intensive networks
    > Solved by adding a penalty in resetLock, as if withdrawal is not fulfilled one has to resetLock, and by penalty, the attacker is stopped from doing it repeatedly.

# Part 3: Dividing StakeManger into 2 contracts

When we added all of the above functionalities to Stakemanager we found we are past the maximum allowed limit - 24kb by a small margin.
Now we could have made a run of the optimizer to a smaller number to come under the allowed size, but we went with dividing the contract itself into 2 contracts, so in the future, we don't face the same issue again.

So old StakeManager is divided into 2 contarcts

1) **RewardManager : Responsible for rewards and penalties**

2) **StakeManager : Responsible mainly for Stake, Unstake, Withdraw, ResetLock**

Access control is modified accordingly as well

RZR Access Control (3)

---

Jun 10, 2021

### CODE REVIEW

**Comment by abhishekvispute** [comment]
> # Delegation of Stake
> ### Fixes #45
> # Part 1: Delegation
> ### What is it
> Delegating responsibility of voting in the network to existing staker, rather than running node yourself.
> 
> ### Why we need it
> Many times general users want to participate in the network but don't have the resources to set up nodes. With this feature, they can participate in the network by delegating their stake to the staker, letting the staker vote on behalf of them, and being part of a reward mechanism.
> 
> ### How we did it
> In Short,
> 
> * `Staker Accepts Delegation`
>   Staker has an option if they want to accept delegation or not
>   If they do, they set up a commission which acts as a fee to the delegator
> * `Delegation`
>   Now delegator sees all stakers who have accepted delegation
>   and choose any, we would call this, **"Chosen one"** from now on
>   Delegators' RZR tokens would be transferred to the stakemanager.
>   Chosen one's stake would be increased
>   X number of sRZR would be minted and transferred to the delegator's account.
>   X would depend on the current valuation of sRZR
> * `Trading of sRZRs`
>   Now delegator can keep them or trade we are prone to that
>   If they trade, the receiver becomes delegator by definition
>   If chosen staker is on reward positive 1 sRZR > 1 RZR, the delegator is at profit.
>   If chosen staker is on reward negative 1 sRZR < 1 RZR, the delegator is at loss.
> * `Withdraw`
>   If the delegator wants to withdraw
>   His sRZRs would be burned and he would get RZRs back at the current valuation.
>   Commission would be paid to the staker.
> 
> To be consistent
> Now even if staker stakes he would get sTokens back.
> 
> Please refer to this sheet for example
> 
> [sRZR](https://www.notion.so/b603abc874b642c4a35ae1db0447e1ce)
> 
> # Part 2: Lock Mechanism
> (If you are not clear why we need lock on delegators, please refer to notion doc)
> 
> As we don't have any storage on delegators we can't go by how we did for stakers in the past.
> even if we create storage then we would have to define separate functions for stakers and delegators.
> **We needed a user agnostic lock mechanism that would work for all**
> That's what Improved lock mechanism is
> We would have a new structure called Lock
> and map it to [address][tokenaddress]
> 
> ```js
> struct Lock {
> uint256 amount; //amount in sTokens
> uint256 withdrawAfter;
> }
> 
> mapping (address=> mapping(address=>Structs.Lock)) public locks; //in StakeStorage
> ```
> 
> In Unstake, the user will declare how much they want to withdraw (in sRZR)
> 
> and they wait for withdraw after period.
> 
> and then within WithDrawReleasePeriod, they withdraw, they will get RZRs back at the current valuation w.r.t sRZR,
> We burn sRZRs, update stakers stake and pass the commission to staker if withdrawal is initiated by the delegator.
> 
> Locks 2 (1)
> 
> Now what is this **"WithdrawReleasePeriod"** we didn't have this before
> ****Why do we this extra restriction?****
> 
> Because without it following case is possible.
> Consider me as a delegator
> I would call unstake even though I don't want to withdraw.
> Now I can withdraw anytime and I would get RZR back at a valuation of withdrawal call.
> 
> If the user fails to withdraw within **WithdrawReleasePeriod**, their lock expires and they have to call resetLock(), resetLock() is penalized to avoid front-run recurring unstake call. Please refer to the following for more details
> 
> **Attacks**
> 
> 1. **P+E**
>    If the attacker can use influence to signal in-opposite and influenced users are able to leave the network at that moment only without waiting to check if the attacker's signal was true.
>    > Could be solved by forcing staker to participate in network even after unstake
>    > We are not resistant to this with a new design as either we  can resist this or Hit and Run Before Dispute, and we prefer Hit and Run Before Dispute as it's in protocol
> 2. **Extra Protocol Hit and Run**
>    If after unstake attacker can issue negative comments about the network and remain unaffected in withdraw.
>    > Solved by releasing funds as per valuation in withdraw and not in unstake
> 3. **Hit and Run Before Dispute**
>    If the attacker can hit the network and leave before a dispute happens
>    > Solved by locking funds and forcing staker not to participate
> 4. **Front-Run Unstake Call**
>    If the attacker can call unstake though they don't want to withdraw and withdraw anytime after withdraw after period is passed
>    > Solved by the introduction of WithDrawReleasePeriod, the user has to withdraw within this period only, so attackers cannot pre-call unstake, as if they do, after WithDrawReleasePeriod their withdrawal req becomes invalid.
>    
>    
>    **Front-Run Recurring Unstake call**
>    To solve the previous attack we added WithDrawReleasePeriod, but with that as well one can call unstake again each time after WithDrawReleasePeriod is over, on less gas-intensive networks
>    > Solved by adding a penalty in resetLock, as if withdrawal is not fulfilled one has to resetLock, and by penalty, their attacker is stopped from doing it repeatedly.
> 
> # Part 3: Dividing StakeManger into 2 contracts
> When we added all of the above functionalities to Stakemanager we found we are past the maximum allowed limit - 24kb by a small margin.
> Now we could have made a run of the optimizer to a smaller number to come under the allowed size, but we went with dividing the contract itself into 2 contracts, so in the future, we don't face the same issue again.
> 
> So old StakeManager is divided into 2 contarcts
> 
> 1. **StakeRegulator : Responsible for rewards and penalties**
> 2. **StakeManager : Responsible mainly for Stake, Unstake, Withdraw, ResetLock**
> 
> Access control is modified accordingly as well
> 
> RZR Access Control (1)
> 
> Jun 10, 2021

File: contracts/Core/StakeManager.sol [hunk 1]
```diff
@@ -116,15 +109,10 @@ contract StakeManager is Initializable, ACL, StakeStorage {
...
-    function updateBlockReward(uint256 _blockReward) external onlyRole(parameters.getDefaultAdminHash())
-    {
-        blockReward = _blockReward;
-    }
-
     /// @notice stake during commit state only
     /// we check epoch during every transaction to avoid withholding and rebroadcasting attacks
     /// @param epoch The Epoch value for which staker is requesting to stake
-    /// @param amount The amount of schelling tokens Staker stakes
+    /// @param amount The amount in RZR
```
Comment (SkandaBhat) [inline contracts/Core/StakeManager.sol:115]: Ideally this must be renamed in a separate PR to address #156. Also, RAZOR*

File: contracts/Core/Parameters.sol [hunk 2]
```diff
@@ -128,4 +139,12 @@ contract Parameters is IParameters, ACL {
...
-}
+
+    function getStakeRegulatorHash() external pure override returns (bytes32) {
+        return _STAKE_REGULATOR_HASH;
+    }
+
+    function getRewardPoolModifierHash() external pure override returns (bytes32) {
+        return _REWARD_POOL_MODIFER_HASH;
+    }
+}
```
Comment (SkandaBhat) [inline contracts/Core/Parameters.sol:150]: It is generally a good practise to end files with new line - https://stackoverflow.com/a/729795

File: test/helpers/testSetup.js [hunk 3]
```diff
@@ -64,9 +71,11 @@ const setupContracts = async () => {
     random,
     schellingCoin,
     stakeManager,
+    stakeRegulator,
```
Comment (SkandaBhat) [inline test/helpers/testSetup.js:74]: Add this at the end with stakedToken

File: test/StakeManager.js [hunk 4]
```diff
@@ -21,17 +30,23 @@ describe('StakeManager', function () {
...
     let initializeContracts;
+    let stakedToken;
+    let random;
 
     before(async () => {
       ({
         schellingCoin,
         blockManager,
         stakeManager,
+        stakeRegulator,
```
Comment (SkandaBhat) [inline test/StakeManager.js:44]: Add stakeRegulator at the end

File: test/StakeManager.js [hunk 5]
```diff
@@ -21,17 +30,23 @@ describe('StakeManager', function () {
     let blockManager;
     let parameters;
     let stakeManager;
+    let stakeRegulator;
```
Comment (SkandaBhat) [inline test/StakeManager.js:33]: Add stakeRegulator to the end, the reason being it keeps the diff clean and prevents potential conflicts.

I see the reason for adding it here as it makes sense to add it along stakeManager. You can take a call with regards to my suggestion vs how it is right now.

File: package.json [hunk 6]
```diff
@@ -35,6 +35,7 @@
   "dependencies": {
     "@openzeppelin/contracts": "^4.0.0",
     "@razor-network/merkle": "^0.6.0",
+    "hardhat-contract-sizer": "^2.0.3",
```
Comment (SkandaBhat) [inline package.json:38]: Is this required anymore?

File: migrations/src/postDeploymentSetup.js [hunk 7]
```diff
@@ -69,15 +71,19 @@ module.exports = async () => {
     pendingTransactions.push(await schellingCoin.transfer(faucetAddress, SEED_AMOUNT));
   }
 
-  pendingTransactions.push(await blockManager.initialize(stakeManagerAddress, voteManagerAddress, assetManagerAddress, parametersAddress));
-  pendingTransactions.push(await voteManager.initialize(stakeManagerAddress, blockManagerAddress, parametersAddress));
-  pendingTransactions.push(await stakeManager.initialize(schellingCoinAddress, voteManagerAddress, blockManagerAddress, parametersAddress));
+  pendingTransactions
```
Comment (SkandaBhat) [inline migrations/src/postDeploymentSetup.js:74]: Irregular line break

**Review by SkandaBhat — COMMENTED** [review]
Looks good to me, have mentioned a few opinions, please check if they sound fair to you.

File: hardhat.config.js [hunk 8]
```diff
@@ -11,6 +11,7 @@ require('@nomiclabs/hardhat-ethers');
 require('@nomiclabs/hardhat-truffle5');
 require('hardhat-gas-reporter');
 require('solidity-coverage');
+require('hardhat-contract-sizer');
```
[the remark this answers is not in the document]
Comment (abhishekvispute) [reply hardhat.config.js:14]: There is no hard requirement, it just measures contract size after compilation, so we can check if we are under 24kb.
Will remove it, we can add it anytime if needed

File: contracts/Core/StakeManager.sol (hunk 1 above)
Comment (abhishekvispute) [reply contracts/Core/StakeManager.sol:115]: Sure will make this change

File: contracts/Core/Parameters.sol (hunk 2 above)
Comment (abhishekvispute) [reply contracts/Core/Parameters.sol:150]: Yep will do

File: test/helpers/testSetup.js (hunk 3 above)
Comment (abhishekvispute) [reply test/helpers/testSetup.js:74]: Hmm I think its better with stakeManager as both are related
And as of now I am not sure if we are going with this naming itself.
Lets see this once we all agree that stakeRegulator and stakeManager are final names.

File: hardhat.config.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhishekvispute) [reply hardhat.config.js]: Done https://github.com/razor-network/contracts/pull/121/commits/e61016b7ecbf64a54dbbd3db02332c2addcd06af

File: contracts/Core/Parameters.sol (hunk 2 above)
Comment (abhishekvispute) [reply contracts/Core/Parameters.sol]: Done https://github.com/razor-network/contracts/pull/121/commits/e61016b7ecbf64a54dbbd3db02332c2addcd06af

File: package.json (hunk 6 above)
Comment (abhishekvispute) [reply package.json]: Done https://github.com/razor-network/contracts/pull/121/commits/e61016b7ecbf64a54dbbd3db02332c2addcd06af

File: migrations/src/postDeploymentSetup.js (hunk 7 above)
Comment (abhishekvispute) [reply migrations/src/postDeploymentSetup.js]: Done https://github.com/razor-network/contracts/pull/121/commits/e61016b7ecbf64a54dbbd3db02332c2addcd06af

File: contracts/Core/StakeManager.sol (hunk 1 above)
Comment (SkandaBhat) [reply contracts/Core/StakeManager.sol:115]: makes sense. Lets formalise it while addressing #156.

**Review by SkandaBhat — APPROVED** [review]
LGTM

**Review by SamAg19 — DISMISSED** [review]

**Comment by SamAg19** [comment]
LGTM

**Review by hrishikeshio — APPROVED** [review]

**Comment by hrishikeshio** [comment]
LGTM

**Review by SamAg19 — APPROVED** [review]

**Comment by SamAg19** [comment]
LGTM
